### PR TITLE
xdc conflict resolution

### PR DESCRIPTION
**Pull request by yux0**
#1457 Conflict resolve algorithm

### CODE REVIEW

File: common/xdc/eventReplication.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	return &eventBranch{
+		branchToken: token,
+		versionHistory: eventNodes,
+	}
+}
+
+// add is to add new node into the versionHistory slice
+func (e *eventBranch) add(node eventNode) {
+	if node.eventID <= e.getLastNode().eventID || node.version <= e.getLastNode().version {
+		// Or panic?
```
Comment (wxing1292) [inline common/xdc/eventReplication.go]: it is better to panic

File: common/xdc/eventReplication.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+func (e *eventBranch) updateEventID(eventID int64) {
+	if eventID <= e.getLastNode().eventID {
+		// Or panic?
+		return
+	}
+	e.getLastNode().eventID = eventID
+}
+
+// append is to add or update the versionHistory slice
+func (e *eventBranch) append(node eventNode) {
```
Comment (wxing1292) [inline common/xdc/eventReplication.go]: this function can be merged with updateEventID and add

File: common/xdc/eventReplication.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		if eID > eventID {
+			eventID = eID
+			branchIndex = i
+		}
+	}
+	return
+}
+
+func (m *eventReplicationMetadata) findMinEventIDWithMaxSameVersion(local, remote eventBranch) int64 {
+	localIdx := len(local.versionHistory) - 1
```
Comment (wxing1292) [inline common/xdc/eventReplication.go]: how about this
```
length := len(local.versionHistory)
if length > len(remote.versionHistory) {
	length = len(remote.versionHistory)
}

index := -1
for i := 0; i < length; i++ {
	if local.versionHistory[i].version == remote.versionHistory[i].version {
		index = i
	}
}

if index < 0 {
	panic("version history is malformed")
}

eventID := common.MinInt64(local.versionHistory[index].eventID, remote.versionHistory[index].eventID)
version := local.versionHistory[index].version
return eventNode{
	version: version,
	eventID: eventID,
}
```

File: common/xdc/eventReplication.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+func newEventReplicationMetadata(branches ...eventBranch) *eventReplicationMetadata {
+	return &eventReplicationMetadata {
+		eventsBranches: branches,
+	}
+}
+
+func (m *eventReplicationMetadata) merge(branch eventBranch) {
+	eventID, branchIdx := m.findBranchWithMaxEventID(branch)
+
+	if m.eventsBranches[branchIdx].versionHistory[len(m.eventsBranches[branchIdx].versionHistory) - 1].eventID == eventID {
```
Comment (wxing1292) [inline common/xdc/eventReplication.go]: it is possible that the incoming branch contains more information, i.e. during backfill, the version history of a replication task can be in the future

File: common/xdc/eventReplication.go (hunk 1 above)
Comment (yux0) [reply common/xdc/eventReplication.go]: Updated

File: common/xdc/eventReplication.go (hunk 2 above)
Comment (yux0) [reply common/xdc/eventReplication.go]: Merge those three func to one func

File: common/xdc/eventReplication.go (hunk 3 above)
Comment (yux0) [reply common/xdc/eventReplication.go]: Which one do you prefer? This is top down search and mine is bottom up ?

File: common/persistence/versionHistoryStore.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	versionHistoriesImpl struct {
+		versionHistories []VersionHistory
+	}
+)
+
+// NewVersionHistory initializes new version history
+func NewVersionHistory(items ...VersionHistoryItem) VersionHistory {
+	return &versionHistoryImpl{
```
Comment (wxing1292) [inline common/persistence/versionHistoryStore.go]: plz add a validation here

File: common/persistence/versionHistoryStore.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+		history: items,
+	}
+}
+
+// Update updates the versionHistory slice
+func (v *versionHistoryImpl) Update(item VersionHistoryItem) error {
+	if item.version < v.getLastItem().version {
+		return &shared.BadRequestError{
+			Message: fmt.Sprintf("cannot update version history with a lower version %v. Current version: %v",
```
Comment (wxing1292) [inline common/persistence/versionHistoryStore.go]: `Current version` -> `last version`

File: common/persistence/versionHistoryStore.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+				item.version,
+				v.getLastItem().version),
+		}
+	}
+
+	if item.eventID < v.getLastItem().eventID {
```
Comment (wxing1292) [inline common/persistence/versionHistoryStore.go]: <=, not <

File: common/persistence/versionHistoryStore.go (hunk 5 above)
Comment (wxing1292) [reply common/persistence/versionHistoryStore.go]: including size checking, making sure len(history) > 0

File: common/persistence/versionHistoryStore.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+		}
+	}
+	return VersionHistoryItem{}, &shared.BadRequestError{
+		Message: fmt.Sprintf("version history is malformed. No joint point found."),
+	}
+}
+
+// IsAppendable checks if a version history item is appendable
+func (v *versionHistoryImpl) IsAppendable(item VersionHistoryItem) bool {
+	return v.getLastItem().eventID == item.eventID && v.getLastItem().version == item.version
```
Comment (wxing1292) [inline common/persistence/versionHistoryStore.go]: `v.getLastItem() == item`

File: common/persistence/versionHistoryStore.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+// NewVersionHistories initialize new version histories
+func NewVersionHistories(histories ...VersionHistory) VersionHistories {
+	return &versionHistoriesImpl{
+		versionHistories: histories,
+	}
+}
+
+// FindLowestCommonVersionHistory finds the lowest common version history item among all version histories
+func (h *versionHistoriesImpl) FindLowestCommonVersionHistory(history VersionHistory) (VersionHistoryItem, *VersionHistory, error) {
+	var hItem VersionHistoryItem
```
Comment (wxing1292) [inline common/persistence/versionHistoryStore.go]: versionHistoryItem

File: common/persistence/versionHistoryStore.go [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+	var vHistory *VersionHistory
```
Comment (wxing1292) [inline common/persistence/versionHistoryStore.go]: versionHistory

File: common/persistence/versionHistoryStore.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+	for i := 0; i < len(h.versionHistories); i++ {
```
Comment (wxing1292) [inline common/persistence/versionHistoryStore.go]:
```
for _, versionHistory := range h.versionHistories {
	item, err := versionHistory.FindLowestCommonVersionHistoryItem(history)
	if err != nil {
		return err
	}

	if item.eventID > hItem.eventID {
		versionHistoryItem = item
		versionHistory = versionHistory
	}
}
```

File: common/persistence/versionHistoryStore.go (hunk 10 above)
Comment (wxing1292) [reply common/persistence/versionHistoryStore.go]: please do not use pointer to interface

File: common/persistence/versionHistoryStore.go (hunk 5 above)
Comment (yux0) [reply common/persistence/versionHistoryStore.go]: Updated

File: common/persistence/versionHistoryStore.go (hunk 6 above)
Comment (yux0) [reply common/persistence/versionHistoryStore.go]: updated

File: common/persistence/versionHistoryStore.go (hunk 7 above)
Comment (yux0) [reply common/persistence/versionHistoryStore.go]: updated

File: common/persistence/versionHistoryStore.go (hunk 8 above)
Comment (yux0) [reply common/persistence/versionHistoryStore.go]: updatedd

File: common/persistence/versionHistoryStore.go (hunk 9 above)
Comment (yux0) [reply common/persistence/versionHistoryStore.go]: updated

File: common/persistence/versionHistoryStore.go (hunk 10 above)
Comment (yux0) [reply common/persistence/versionHistoryStore.go]: updated

File: common/persistence/versionHistoryStore.go (hunk 11 above)
Comment (yux0) [reply common/persistence/versionHistoryStore.go]: updated

File: common/persistence/versionHistory.go [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+	copy(history, items)
+	return VersionHistory{
+		history: history,
+	}
+}
+
+// Update updates the versionHistory slice
+func (v *VersionHistory) Update(item VersionHistoryItem) error {
+	currentItem := item
+	lastItem := &v.history[len(v.history)-1]
```
Comment (wxing1292) [inline common/persistence/versionHistory.go]: why '&'?

File: common/persistence/versionHistory.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+// IsAppendable checks if a version history item is appendable
+func (v *VersionHistory) IsAppendable(item VersionHistoryItem) bool {
+	return v.history[len(v.history)-1] == item
+}
+
+// NewVersionHistories initialize new version histories
+func NewVersionHistories(histories []VersionHistory) VersionHistories {
+	if histories == nil || len(histories) == 0 {
```
Comment (wxing1292) [inline common/persistence/versionHistory.go]: length check will be enough

File: common/persistence/versionHistory.go [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	return versionHistoryItem, versionHistory, nil
+}
+
+// AddHistory add new history into version histories
+// Sample
+func (h *VersionHistories) AddHistory(item VersionHistoryItem, local VersionHistory, remote VersionHistory) error {
+	commonItem := item
+	if local.IsAppendable(commonItem) {
+		for _, historyItem := range remote.history {
```
Comment (wxing1292) [inline common/persistence/versionHistory.go]: after this for loop, is local == remote?

File: common/persistence/versionHistory.go [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+	}
+)
+
+// NewVersionHistory initializes new version history
+func NewVersionHistory(items []VersionHistoryItem) VersionHistory {
+	if len(items) == 0 {
+		panic("version history items cannot be empty")
+	}
+
+	history := make([]VersionHistoryItem, len(items))
```
Comment (wxing1292) [inline common/persistence/versionHistory.go]: not necessary to make a deep copy here

Comment (yux0) [reply common/persistence/versionHistory.go]: Yes. Removed it

File: common/persistence/versionHistory.go (hunk 14 above)
Comment (yux0) [reply common/persistence/versionHistory.go]: Correct. I will remove this loop. But there will be loop to find the passed in versionHistory in the versionHistories. After the two functions merged, this loop will not be needed

File: common/persistence/versionHistory.go (hunk 13 above)
Comment (yux0) [reply common/persistence/versionHistory.go]: done

**Pull request merged**